source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,300 @@ namespace EnvoyDefault {
-  return std::make_unique<HeaderValidator>(config_, protocol, stream_info);
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidMethod};
+  }
+
+  return HeaderEntryValidationResult::success();
 }
 
-HeaderValidator::HeaderValidator(const HeaderValidatorConfig&, ::Envoy::Http::Protocol,
-                                 StreamInfo::StreamInfo& stream_info)
-    : stream_info_(stream_info) {}
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateSchemeHeader(const HeaderString& value) {
+  // From RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.1:
+  //
+  // scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+  //
+  // Although schemes are case-insensitive, the canonical form is lowercase and documents that
+  // specify schemes must do so with lowercase letters. An implementation should accept uppercase
+  // letters as equivalent to lowercase in scheme names (e.g., allow "HTTP" as well as "http") for
+  // the sake of robustness but should only produce lowercase scheme names for consistency.
+  //
+  // The validation mode controls whether uppercase letters are permitted.
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+  }
+
+  auto character_it = value_string_view.begin();
+
+  // The first character must be an ALPHA
+  auto valid_first_character = (*character_it >= 'a' && *character_it <= 'z') ||
+                               (*character_it >= 'A' && *character_it <= 'Z');
+  if (!valid_first_character) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+  }
+
+  for (++character_it; character_it != value_string_view.end(); ++character_it) {
+    if (!testChar(kSchemeHeaderCharTable, *character_it)) {
+      return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidScheme};
+    }
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateStatusHeader(const HeaderString& value) {
+  // This is based on RFC 7231, https://datatracker.ietf.org/doc/html/rfc7231#section-6,
+  // describing the list of response status codes and the list of registered response status codes,
+  // https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml.
+
+  static uint32_t kMinimumResponseStatusCode = 100;
+  static uint32_t kMaximumResponseStatusCode = 999;
+  const auto& value_string_view = value.getStringView();
+  auto buffer_end = value_string_view.data() + value_string_view.size();
+
+  // Convert the status to an integer.
+  std::uint32_t status_value{};
+  auto result = std::from_chars(value_string_view.begin(), value_string_view.end(), status_value);
+  if (result.ec == std::errc::invalid_argument || result.ptr != buffer_end) {
```

Comment:
Nit: 
```suggestion
  if (result.ec == std::errc::invalid_argument || result.ptr != value_string_view.end()) {
```

And kill the `buffer_end` above. There is similar code below too.